text/3700-ptr-assume-moved.md

```diff
@@ -0,0 +1,257 @@
+high-bit tagging would have to resort to manually using bitwise operations and would be at risk of
+inadvertently introducing Undefined Behaviour.  Having a helper for doing so in the library creates
+a place where e.g. Miri hooks can be called to let Miri know that a pointer's cannonical address has
+been updated. It also makes it possible for the helper to receive special treatment in the codegen
+stage, which is required for it to have the desired behaviour.
+
+It is most likely not feasible to make `assume_moved()` safe to use regardless of the context,
+hence the current approach is to make it an unsafe function with a safety notice about the user's
+responsibilities.
+
+# Prior art
+[prior-art]: #prior-art
+
+TBI already works in C, though mostly by default and care must be taken to make sure no Undefined
+Behaviour is introduced. The compiler does not take special steps to preserve the tags, but it
+doesn't try to remove them either.  That being said, the C/C++ standard library does not take
+tagging schemes into account during alias analysis. With this proposal, Rust would have much better
+defined and safer support for TBI than C or C++.
+
+Notably, [Android](https://source.android.com/docs/security/test/tagged-pointers) already makes
+extensive use of TBI by tagging all heap allocations.
+
+The idea is also not one specific to AArch64, as there are similar extensions present on other
+architectures that facilitate working with tagged pointers.
+This proposal is in no way AArch64-specific. As proposed, `assume_moved` could be used on x86
+for Intel's LAM or AMD's UAI as well. It could even be used for applications other than pointer
+tagging, such as using `mmap` to map the same allocation to two adjacent addresses and switch between
+them by using `assume_moved()`. Although that use-case requires its own documentation and
+additional concurrency safety considerations that do not apply to pointer tagging.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+What exactly should the helper be called? The main candidates so far have been `simulate_realloc`,
+`simulate_move` and `assume_moved`, alongside some other suggestions, but preferences seem to be
+rather subjective and no clearly most suitable name has emerged as of yet. It can be renamed to
+whichever name the relevant Rust team finds most appropriate.
+
+What is the best way to make this compatible with Strict Provenance? We want to be able to create a
+pointer with an arbitrary address, detached from any existing pointers and with a brand-new
+provenance. From the LLVM side this can be handled through generating `inttoptr` which does not have
+the same aliasing restrictions as `getelementptr` alongside annotating the function return value as
+`noalias` which can be done with the aforementioned new built-in attribute. Is this enough for it to
+fit within the Strict Provenance framework? If not, how can we make it fit?
```

Comment:
@RalfJung, what are your thoughts on this?  Do we need to leave this as an open question, or do we have any answers here?  It would seem a substantial open question to leave.

Reply:
I can't follow the arguments in the text as all -- what it proposes for LLVM makes little sense to me. I'd like to hear from @nikic what the best way would be to model this for LLVM; my proposal would be an inline asm block.

From the Rust side, we can just say that this is a `realloc` with a chosen address (and it is UB if that would make the allocation overlap with some other allocation). I don't know which potential problems with strict provenance the RFC refers to.

Reply:
Which part does not make sense? The implementation needs to do 3 things:
1. Create the pointer
2. "Launder" it through a black box of some kind to prevent incorrect optimisations
3. Make the pointer be treated as if it were returned from an allocator function

For which the solutions I'm suggesting are:
1. `inttoptr`, because `getelementptr` is UB in this scenario
2. inline asm block
3. `noalias` return annotation

We need all 3 of the above for this to actually behave as described. At the very least I've not seen a different way of doing it.

Reply:
> It’s invalid to take a GEP from one object, address into a different separately allocated object, **and dereference it**.

Emphasis mine. You're not dereferencing the result of the GEP. You're passing it into the laundromat. There's only two options now:
- laundry worked, LLVM lost track of the pointer, there's no problem
- laundry failed, LLVM recognizes this is the old pointer and everything is broken no matter how many `inttoptr` you add.

> I think it's easier to just use inttoptr and sidestep this entirely.

I think it is bad to be cautious without an understanding of what we'd be cautioning against. And once you dig into it and understand what happens here, it's just pointless and confusing.